### PR TITLE
feat(affiliate): improve order checking for both eligiblity and ref code recovery

### PR DESCRIPTION
**Pull request by kernelwhisperer**
# Summary

Fixes [Edit a referral code after expired/cancelled trade](https://www.notion.so/cownation/Edit-a-referral-code-after-expired-cancelled-trade-3138da5f04ca80a8a089c96cbbfcb86b)

1. We should only check for executed & non-integrator orders using `sdk.getOrders` ( `sdk.getTrades` does not return app data) for both **eligiblity** and **ref code recovery**
2. Improves the dune dashboard sql to account for these edge cases:
    - Binding is fixed to the first non-integrator trade with a `referrer_code`.
    - Trades with a different code after binding are ignored for rewards.
    - Integrator trades are ignored both before and after binding: they do not bind code, do not affect new-trader eligibility, and do not add eligible volume.
    - New-trader rule is strict on non-integrator trades: if first non-integrator trade has no code, later coded trades are ineligible.

- 3. Now that `checkIfTraderHasPastTrades` is almost identical to `findRefCodeInPastTrades`, I will merged them into a single function that returns `{ hasPastTrades, refCode }` to avoid extra api calls. (dc6946a8da5c4874d19e59a17313d5fd6652d0a0)

> [!TIP] 
> 
> Review each commit individually.

## Testing

1. Try a integrator trade on a new account: any widget deployment, e.g.: Safe
2. Check for eligiblity on the UI

Same for a cancelled/expired order.

I tested with an [account that has a cancelled order](https://explorer.cow.fi/address/0x56fd29500c72CC7Bae2DF0Fb3e4Ac4e150c1f8E7) and still eligible.

## Summary by CodeRabbit

* **New Features**
  * New hook surfaces a trader's past-order eligibility and returns a referral code (if found); eligibility flow updated to use this hook.

* **Bug Fixes**
  * Integrator-sourced trades are excluded from first-trade/eligibility calculations.
  * Cancelled, expired, or zero-filled orders are ignored when assessing past activity.

* **Documentation**
  * Added explicit edge-case rules for binding, integrator interactions, and reward eligibility.

* **Tests**
  * Expanded coverage for past-trade checks, multi-environment scanning, and order-filtering scenarios.

### CODE REVIEW

File: apps/cowswap-frontend/src/modules/affiliate/hooks/useRefCodeFromOrderbookTrades.ts [hunk 1]
```diff
@@ -21,9 +18,11 @@ export function useRefCodeFromOrderbookTrades(
...
   const { account, enabled } = params
 
-  return useSWR<string | undefined>(
-    enabled && !!account ? ['affiliate-refCode-orderbook', account] : null,
-    async () => (!account ? undefined : findRefCodeInPastTrades(account)),
-    { refreshInterval: AFFILIATE_ORDERBOOK_REFRESH_INTERVAL_MS },
-  )
+  const { data, isLoading, error } = useAffiliateTraderPastOrders({ account, enabled })
+
+  return {
```
Comment (shoom3301) [inline apps/cowswap-frontend/src/modules/affiliate/hooks/useRefCodeFromOrderbookTrades.ts]: Can we just do?:
```
return useAffiliateTraderPastOrders({ account, enabled })
```

I'm asking becuse the hook returns literal object and we agreed to use useMemo for such cases

File: apps/cowswap-frontend/src/modules/affiliate/hooks/useHasOrderbookTrades.ts [hunk 2]
```diff
@@ -22,16 +19,11 @@ interface UseHasOrderbookTradesResult {
...
-      revalidateOnReconnect: false,
-      dedupingInterval: AFFILIATE_ORDERBOOK_REFRESH_INTERVAL_MS,
-      refreshInterval: AFFILIATE_ORDERBOOK_REFRESH_INTERVAL_MS,
-      revalidateIfStale: true,
-      revalidateOnMount: true,
-    },
-  )
+  const { data, isLoading, error } = useAffiliateTraderPastOrders({ account, enabled })
+
+  return {
```
Comment (shoom3301) [inline apps/cowswap-frontend/src/modules/affiliate/hooks/useHasOrderbookTrades.ts]: Same as https://github.com/cowprotocol/cowswap/pull/7111/changes#r2894536214

File: apps/cowswap-frontend/src/modules/affiliate/hooks/useAffiliateTraderPastOrders.ts [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+  return useSWR<PastTradesCheckResult>(
+    enabled && !!account ? ['affiliate-orderbook-trades-check', account] : null,
+    async () =>
+      !account
+        ? {
+            hasPastTrades: false,
+            refCode: undefined,
+          }
+        : checkIfTraderHasPastTrades(account),
+    {
```
Comment (shoom3301) [inline apps/cowswap-frontend/src/modules/affiliate/hooks/useAffiliateTraderPastOrders.ts:26]: It's better to move the object out of the hook scope and define it as const on the file level

File: apps/cowswap-frontend/src/modules/affiliate/api/checkIfTraderHasPastTrades.ts [hunk 4]
```diff
@@ -1,21 +1,39 @@
...
-export async function checkIfTraderHasPastTrades(owner: Address): Promise<boolean> {
+/**
+ * Scans executed non-integrator orders across supported chain/env pairs and returns whether
+ * the trader has past orders, which can be used to determine trader eligibility, plus the
+ * first valid affiliate ref code found in their appData.
+ */
+export async function checkIfTraderHasPastTrades(owner: Address): Promise<PastTradesCheckResult> {
   logAffiliate(safeShortenAddress(owner), 'Calling cow api for each supported chain/env to check for executed trades')
 
   const tradesByChainAndEnv = await Promise.all(
```
Comment (shoom3301) [inline apps/cowswap-frontend/src/modules/affiliate/api/checkIfTraderHasPastTrades.ts]: Nitpick:

```suggestion
  const ordersByChainAndEnv = await Promise.all(
```

Comment (kernelwhisperer) [reply apps/cowswap-frontend/src/modules/affiliate/api/checkIfTraderHasPastTrades.ts]: It's a relic of the past for sure, the name of the function is also `checkIfTraderHasPastTrades`, I kept it like this for a minimal diff / less review burden, ~but I will add a new commit to rename the fn + file too...~
Fixed this in a new commit but I kept the function/file name: trade terminology is too ingrained in the current implementation and orderbookOrders is pretty bad name...

File: apps/cowswap-frontend/src/modules/affiliate/hooks/useRefCodeFromOrderbookTrades.ts (hunk 1 above)
Comment (kernelwhisperer) [reply apps/cowswap-frontend/src/modules/affiliate/hooks/useRefCodeFromOrderbookTrades.ts]: I kept these hooks for minimal diff, but in this case, I will just remove both hooks and rely on useAffiliateTraderPastOrders directly

Comment (kernelwhisperer) [reply apps/cowswap-frontend/src/modules/affiliate/hooks/useRefCodeFromOrderbookTrades.ts]: Removed this file in ce986988a1e8695cfe7b7a0d931db52016665a73 now we use `useAffiliateTraderPastOrders` directly.

File: apps/cowswap-frontend/src/modules/affiliate/api/checkIfTraderHasPastTrades.ts [hunk 5]
```diff
@@ -1,21 +1,39 @@
...
   logAffiliate(safeShortenAddress(owner), 'Calling cow api for each supported chain/env to check for executed trades')
 
-  const tradesByChainAndEnv = await Promise.all(
+  const ordersByChainAndEnv = await Promise.all(
     AFFILIATE_SUPPORTED_CHAIN_IDS.flatMap((chainId) =>
       TRADE_ENVS_TO_CHECK.map(async (env) => {
         try {
-          return await getTrades({ owner, limit: 1 }, { chainId, env })
+          const orders = await getOrders({ owner, limit: PAST_ORDERS_SCAN_LIMIT }, { chainId, env })
+          return orders.filter(isExecutedNonIntegratorOrder)
```
[the remark this answers is not in the document]
Comment (kernelwhisperer) [reply apps/cowswap-frontend/src/modules/affiliate/api/checkIfTraderHasPastTrades.ts:36]: No, Safe app is an exception, it actually means a safe wallet using our ui, different from safe widget integration where the app code is "Safe Wallet Swaps"

File: apps/cowswap-frontend/src/modules/affiliate/api/checkIfTraderHasPastTrades.ts [hunk 6]
```diff
@@ -28,10 +46,20 @@ export async function checkIfTraderHasPastTrades(owner: Address): Promise<boolea
     ),
   )
 
-  const hasPastTrades = tradesByChainAndEnv.some((trades) => trades.length > 0)
+  const hasPastTrades = ordersByChainAndEnv.some((orders) => orders.length > 0)
+  const refCode = ordersByChainAndEnv
+    .flat()
+    .map((order) => getRefCodeFromAppData(extractFullAppDataFromOrder(order)))
+    .find((code) => !!code)
```
[the remark this answers is not in the document]
Comment (kernelwhisperer) [reply apps/cowswap-frontend/src/modules/affiliate/api/checkIfTraderHasPastTrades.ts:53]: That's fine, the code should be the same. The only they diverge is if someone bypasses our UI but we don't care about that path
